### ISSUE TITLE
Consider changing the input of `TinyLexer`

### ISSUE THREAD

**Issue by Possseidon** [Possseidon/parser-playground#6]
Currently, `TinyLexer` takes a string slice `code: &str` as input. While this makes it easy to work with in the lexer, it's not actually necessary, since the returned tokens do not reference into the original input (they are returned as `SmolStr`).

If `TinyLexer` instead took some form of [`Read`](https://doc.rust-lang.org/std/io/trait.Read.html), it could then operate directly on e.g. a [`File`](https://doc.rust-lang.org/std/fs/struct.File.html) handle without keeping the entire file in memory. [`Seek`](https://doc.rust-lang.org/std/io/trait.Seek.html) might also be helpful, although I think it's better if I try and avoid that.

Basically I want to replace `code: &str` in `TinyLexer` with a generic `T: Read`.

`PositionedLexer` can't really benefit from this, since tokens reference into the original string.

**Comment by Possseidon** [comment]
Thinking about it more and more, it's probably not worth the effort. The whole scanning char by char to find e.g. idents would be really slow with an unbuffered input. And even if you do pass in a BufReader it would still be better to just have the full thing in memory consecutively.
